pkg/controller/controller.go

```diff
@@ -560,29 +560,46 @@ func (c *M3DBController) handleClusterUpdate(cluster *myspec.M3DBCluster) error
 			return err
 		}
 
-		if !update {
+		// If we're not updating the statefulset AND we're not using the OnDelete update
+		// strategy, then move to the next statefulset. When using the OnDelete update
+		// strategy, we still may want to restart nodes for this particular statefulset,
+		// so don't continue yet.
+		if !update && !cluster.Spec.OnDeleteUpdateStrategy {
 			continue
 		}
 
-		_, err = c.kubeClient.AppsV1().StatefulSets(cluster.Namespace).Update(expected)
-		if err != nil {
-			c.logger.Error(err.Error())
-			return err
+		// processNext indicates whether we should process the next statefulset in the loop.
+		// processNext will be true if the statefulset or pods for this statefulset
+		// are not updated. We do not process the next statefulset if we've performed any
+		// updates as we want to wait for those updates to be completed successfully first
+		// (i.e. new pods are in the ready state).
+		processNext := true
+		if update {
+			actual, err = c.applyStatefulSetUpdate(cluster, actual, expected)
+			if err != nil {
+				c.logger.Error(err.Error())
+				return err
+			}
+			processNext = false
```

Comment:
Rather than have `processNext = false` here and continue this loop, can we just `return nil` so that we exit + re-trigger the entire reconciliation loop? That way we'll run all of our safety checks again.

Reply:
AFAICT this would also match the old post-update behavior.

Reply:
lol, yes, that's much simpler. Will update.

Reply:
Actually, wait. Only, if the updateStrategy is rolling update can we do that. Still probably nice to bail early if using rolling update, so will modify for that.

Reply:
OK, pushed a change to clarify this logic even more. It's worth noting that none of the semantics around this loop have changed here. In both the RollingUpdate and OnDelete cases, if changes are made to a statefulset, the operator exits this loop and relies on event updates (i.e. pod ready) to continue updating.